src/hardware/mixer.cpp

```diff
@@ -2540,146 +2518,86 @@ static void SDLCALL mixer_callback([[maybe_unused]] void* userdata,
+{
+	double last_mixed = 0.0;
+	while (!mixer.thread_should_quit) {
+		std::unique_lock lock(mixer.mutex);
+		assert(mixer.state != MixerState::Uninitialized);
 
-	if (frames_requested != reduce_frames) {
-		// We're doing a very crude sample-skipping style audio
-		// stretching here. However, it's worth keeping it as this path
-		// is almost exclusively used in the fast-forward mode to
-		// achieves that cool "tape speed-up" effect.
-		//
-		// Without this effect, the audio becomes a crackling mess when
-		// fast-forwarding.
-		auto num_frames = std::min(reduce_frames, frames_requested);
-		auto work_pos   = mixer.work.begin() + mixer.pos.load();
+		// This code is mostly for the fast-forward button (hold Alt + F12)
+		const double now = PIC_FullIndex();
+		const double actual_time = now - last_mixed;
+		const double expected_time = (static_cast<double>(mixer.blocksize) / static_cast<double>(mixer.sample_rate_hz)) * 1000.0;
+		last_mixed = now;
 
-		while (num_frames--) {
-			index += index_add;
+		// "Underflow" is not a concern since moving to a threaded mixer
+		// If the CPU is running slower than real-time, the audio drivers will naturally slow down the audio
+		// Therefore, we can always request at least a blocksize worth of audio
+		int frames_requested = mixer.blocksize;
 
-			const auto frame = *(work_pos + static_cast<int>(index));
+		// Be careful is adjusting this value as the actual time will have some fluctuations in a normal run
+		// False postivies will result in an audible "skip"
+		mixer.chipmunk_mode = actual_time > expected_time * 3.0;
 
-			*output++ = clamp_to_int16(frame.left);
-			*output++ = clamp_to_int16(frame.right);
+		if (mixer.chipmunk_mode) {
+			// The emulation is running faster than real-time
+			// Normally only trigged by the fast-forward button (ALT + F12)
+			// We need to consume more frams to "catch up"
+			frames_requested = ifloor(actual_time * get_mixer_frames_per_tick());
+			assert(frames_requested > mixer.blocksize);
 		}
-	} else {
-		auto num_frames = reduce_frames;
-		auto work_pos   = mixer.work.begin() + mixer.pos.load();
-
-		while (num_frames--) {
-			const auto frame = *work_pos++;
 
-			*output++ = clamp_to_int16(frame.left);
-			*output++ = clamp_to_int16(frame.right);
+		mix_samples(frames_requested);
+		assert(mixer.output_buffer.size() == check_cast<size_t>(frames_requested));
+
+		auto& to_mix = mixer.chipmunk_mode ? mixer.squashed_buffer : mixer.output_buffer;
+
+		lock.unlock();
+
+		if (mixer.chipmunk_mode) {
+			// This is "chipmunk mode" meant for fast-forward
+			// It's basic sample skipping to compress a large amount of audio into a single blocksize
+			assert(frames_requested > mixer.blocksize);
+			mixer.squashed_buffer.clear();
+			mixer.squashed_buffer.reserve(mixer.blocksize);
+			const float index_add = static_cast<float>(frames_requested) / static_cast<float>(mixer.blocksize);
+			float float_index = 0.0f;
+			for (int i = 0; i < mixer.blocksize; ++i) {
```

Comment:
minor: some new lines would be good to break up the block of text.

Reply:
Fixed